### ISSUE TITLE
[Bug][move-prover] Boogie compilation error on third-party code

### ISSUE THREAD

**Issue by sblackshear** [diem/diem#9159]
# 🐛 Bug 

Trying to run the prover on some third-party code and encountered a crash

## To reproduce

```
unzip bridge_repro.zip
mvp bridge-repro/bridge_modules --dependency bridge-repro/stdlib_fork
```

**Stack trace/error message**
```
[INFO] translating module Bridge
[INFO] running solver
[internal] boogie exited with compilation errors:
output.bpl(7214,91): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''': $1_Option_Option'$1_Diem_BurnCapability'#0'' (expected: $1_Option_Option'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''')
output.bpl(7223,83): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''': $1_Option_Option'$1_Diem_BurnCapability'#0'' (expected: $1_Option_Option'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''')
output.bpl(7239,148): Error: invalid type for argument 0 in application of $1_Option_$borrow'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''': $1_Option_Option'$1_Diem_BurnCapability'#0'' (expected: $1_Option_Option'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''')
output.bpl(8038,103): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
output.bpl(8076,103): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''': $1_Option_Option'$1_Diem_BurnCapability'#0'' (expected: $1_Option_Option'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''')
output.bpl(9588,103): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
output.bpl(9626,103): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''': $1_Option_Option'$1_Diem_BurnCapability'#0'' (expected: $1_Option_Option'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''')
output.bpl(11089,91): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
output.bpl(11098,83): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
output.bpl(11114,148): Error: invalid type for argument 0 in application of $1_Option_$borrow'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
output.bpl(12266,90): Error: invalid type for argument 0 in application of $IsEqual'$1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0'''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
output.bpl(12276,90): Error: invalid type for argument 0 in application of $IsEqual'$1_Option_Option'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0'''': $1_Option_Option'$1_Diem_BurnCapability'#0'' (expected: $1_Option_Option'$1_Diem_BurnCapability'$1_Diem_BurnCapability'#0''')
output.bpl(14234,91): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
output.bpl(14243,83): Error: invalid type for argument 0 in application of $1_Option_$is_none'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
output.bpl(14259,148): Error: invalid type for argument 0 in application of $1_Option_$borrow'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''': $1_Option_Option'$1_Diem_MintCapability'#0'' (expected: $1_Option_Option'$1_Diem_MintCapability'$1_Diem_MintCapability'#0''')
15 type checking errors detected in output.bpl
```
## System information

Running the prover from recent `diem` hash b1b98fee9679eba37c8d11426e6f53870cdf6306

**Comment by mengxu-fb** [comment]
@sblackshear I am trying to debug it then I realize that I don't have the `bridge_repro.zip`...

**Comment by sblackshear** [comment]
[bridge_repro.zip](https://github.com/diem/diem/files/7193424/bridge_repro.zip)

My bad!

**Comment by mengxu-fb** [comment]
Fixed in #9183.

This bug report as well as issues #9156 and #9155 make me wonder whether it will be a good idea to collect the third-party code and create a smoke test suite for the move prover. But maintaining these code (especially with evolving move language semantics) seem to be hard. Tagging @wrwg @DavidLDill for an opinion.
